R/UploadControls.R

```diff
@@ -0,0 +1,185 @@
+#'
+#' **upload data example with collection_date**
+#'
+#' | position | label | study_subject_id | specimen_type | study_short_code | collection_data |
+#' | ------------- | ----- | ---------------- | ------------- | ---------------- | --------------- |
+#' | A0            | xxx1  | subject_1        | PLASMA        | KAM06            | 2022-04-11      |
+#' | A1            | xxx2  | subject_2        | PLASMA        | KAM06            | 2022-04-11      |
+#'
+#' @param container_name A string specifying the name of the container the samples are in. Names must be unique within each sample type.
+#' @param container_barcode A string specifying the barcode for the container the samples are in. Container barcodes are optional. Barcodes must be unique within each sample type.
+#' @param freezer_address A list specifying the freezer address used to store samples. \cr
+#' Required items in the freezer_address list are `name`, `level_I` and `level_II`.
+#' If the freezer_address type is `minus eighty` then `level_I` and `level_II` items specify the rack and position, respecively.
+#' If the freezer_address type is `minus twenty` then `level_I` and `level_II` items specify the shelf and basket, respecively.
+#' @examples
+#' \dontrun{
+#' UploadSamples(sample_type = "micronix",
+#'               upload_data = tibble(position = c("A0"),
+#'                                    label = c("XXX 1"),
+#'                                    study_subject_id = c("1"),
+#'                                    specimen_type = c("PLASMA"),
+#'                                    study_short_code = c("KAM06"),
+#'                                    collection_date = c("2021-04-10")),
+#'                container_name = "test_container",
+#'                freezer_address = list(name = "TBD",
+#'                                       level_I = "TBD",
+#'                                       level_II = "seve's working basket"))
+#' }
+#' @import dplyr
+#' @import RSQLite
+#' @import lubridate
+#' @export
+
+UploadControls <- function(user_data, control_type, database = Sys.getenv("SDB_PATH")) {
+	tryCatch({
+
+		con <- DBI::dbConnect(RSQLite::SQLite(), database)
+
+		dbBegin(con)
+		now = as.character(lubridate::now())
+		user_data=dplyr::rename(user_data, location_name=name)
+
+		## In the database, controls are stored as "study_subjects". This allows for extractions to be linked back to the original control.
+		## With the current database schema, we need to uniquely identify each study_subject under a single study. Instead of changing this rule,
+		## use the density, strain and percentage to start, and then index the control between [1:count]. If for whatever reason there are
+		## controls that exist under this study already, then "base.count" will be used to offset the indexing so that each appended number
+		## is +1 the second largest number under the study (ie. we always increment by 1).
```

Comment:
I feel like this may be too complicated. If you're adding controls to a study that are identical, then they should be treated as the same subject. It's like adding new specimens for an already existing study subject in the regular sample case